pandas/core/dtypes/dtypes.py

```diff
@@ -811,6 +811,45 @@ def __eq__(self, other: Any) -> bool:
             and tz_compare(self.tz, other.tz)
         )
 
+    def __from_arrow__(
+        self, array: pyarrow.Array | pyarrow.ChunkedArray
+    ) -> DatetimeArray:
+        """
+        Construct DatetimeArray from pyarrow Array/ChunkedArray.
+        """
+        import pyarrow
+        import pyarrow.types
+
+        from pandas.core.arrays import DatetimeArray
+        from pandas.core.arrays.arrow._arrow_utils import (
+            pyarrow_array_to_numpy_and_mask,
+        )
+
+        pa_type = array.type
+        pa_unit = "ns"
+        if pyarrow.types.is_timestamp(pa_type):
+            pa_unit = pa_type.unit
+
+        if isinstance(array, pyarrow.Array):
+            chunks = [array]
+        else:
+            chunks = array.chunks
+
+        results = []
+        for arr in chunks:
+            data, mask = pyarrow_array_to_numpy_and_mask(
```

Comment:
ive been dreading this for a while: isn't this going to cause a problem when the arrow array has non-NA entries with iNaT?

Reply:
Yeah, iNaT integers and NA values will both end up as NaT. Not sure if there's much we can do about that?

Reply:
if we want to be Technically Correct we'd have to check for the presence of iNaTs and warn/raise

Reply:
(or in my preferred world, go back in time and get pyarrow to keep iNaT as a sentinel)

Reply:
That's indeed what pyarrow is doing at the moment: ignoring this, so both get converted to NaN:

```
In [26]: arr = pa.array([np.iinfo("int64").min, None], pa.timestamp('ns'))

In [27]: arr
Out[27]: 
<pyarrow.lib.TimestampArray object at 0x7effe3bea5e0>
[
  1677-09-21 00:12:43.145224192,
  null
]

In [28]: arr.to_numpy(zero_copy_only=False)
Out[28]: array(['NaT', 'NaT'], dtype='datetime64[ns]')
```